test/cpp/qps/client_callback.cc

```diff
@@ -253,18 +253,20 @@ class CallbackStreamingPingPongReactor final
       : client_(client), ctx_(std::move(ctx)), messages_issued_(0) {}
 
   void StartNewRpc() {
-    if (client_->ThreadCompleted()) return;
     ctx_->stub_->experimental_async()->StreamingCall(&(ctx_->context_), this);
     write_time_ = UsageTimer::Now();
     StartWrite(client_->request());
+    writes_done_started_.clear();
     StartCall();
   }
 
   void OnWriteDone(bool ok) override {
-    if (!ok || client_->ThreadCompleted()) {
-      if (!ok) gpr_log(GPR_ERROR, "Error writing RPC");
+    if (!ok) {
+      gpr_log(GPR_ERROR, "Error writing RPC");
+    }
+    if ((!ok || client_->ThreadCompleted()) &&
+        !writes_done_started_.test_and_set()) {
       StartWritesDone();
-      return;
```

Comment:
Is this kosher? skipping the return means we are starting a read after calling writes done

Reply:
Ah nevermind, we need to force a read as the server is waiting on a read before it responds to a writesdone.

Reply:
It's not haraam at all. Yes, that's absolutely allowed since they're on different sides of the stream.

Reply:
(we even have simultaneous Read and WritesDone as one of our tests in the async API). That said, it is important to make sure that we never have more than 1 WritesDone call which is why I protected it with the test_and_set.

Reply:
Right I observed that

Reply:
> It's not haraam at all. Yes, that's absolutely allowed since they're on different sides of the stream.

I see what you did there